pkg/controllers/infracluster/azure.go

```diff
@@ -0,0 +1,205 @@
+	} else if err == nil {
+		return target, nil
+	}
+
+	log.Info(fmt.Sprintf("AzureCluster %s/%s does not exist, creating it", target.Namespace, target.Name))
+
+	apiURL, err := url.Parse(r.Infra.Status.APIServerInternalURL)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse apiUrl: %w", err)
+	}
+
+	port, err := strconv.ParseInt(apiURL.Port(), 10, 32)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse apiUrl port: %w", err)
+	}
+
+	if r.Infra.Status.PlatformStatus == nil {
+		return nil, fmt.Errorf("infrastructure PlatformStatus should not be nil: %w", err)
+	}
+
+	azureClientID, ok := capzManagerBootstrapCredentials.Data["azure_client_id"]
+	if !ok {
+		return nil, fmt.Errorf("failed to get azureClientID: %w", err)
+	}
+
+	azureTenantID, ok := capzManagerBootstrapCredentials.Data["azure_tenant_id"]
+	if !ok {
+		return nil, fmt.Errorf("failed to get azureTenantID: %w", err)
+	}
+
+	providerSpec, err := getAzureMAPIProviderSpec(ctx, r.Client)
+	if err != nil {
+		return nil, fmt.Errorf("error obtaining Azure Provider Spec: %w", err)
+	}
+
+	location, err := r.getAzureLocation(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("error obtaining Azure Cluster location: %w", err)
+	}
+
+	azureClusterIdentity := &azurev1.AzureClusterIdentity{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      r.Infra.Status.InfrastructureName,
+			Namespace: defaultCAPINamespace,
+			Annotations: map[string]string{
+				// The ManagedBy Annotation is set so CAPI infra providers ignore the InfraCluster object,
+				// as that's managed externally, in this case by the cluster-capi-operator's infracluster controller.
+				clusterv1.ManagedByAnnotation: managedByAnnotationValueClusterCAPIOperatorInfraClusterController,
+			},
+		},
+		Spec: azurev1.AzureClusterIdentitySpec{
+			Type:              azurev1.ServicePrincipal,
+			AllowedNamespaces: &azurev1.AllowedNamespaces{NamespaceList: []string{defaultCAPINamespace}},
+			ClientID:          string(azureClientID),
+			TenantID:          string(azureTenantID),
+			ClientSecret:      corev1.SecretReference{Name: clusterSecretName, Namespace: defaultCAPINamespace},
+		},
+	}
+
+	if err := r.Create(ctx, azureClusterIdentity); err != nil && !apierrors.IsAlreadyExists(err) {
+		return nil, fmt.Errorf("failed to create Azure Cluster Identity: %w", err)
+	}
+
+	target = &azurev1.AzureCluster{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      r.Infra.Status.InfrastructureName,
+			Namespace: defaultCAPINamespace,
+			// The ManagedBy Annotation is set so CAPI infra providers ignore the InfraCluster object,
+			// as that's managed externally, in this case by this controller.
+			Annotations: map[string]string{
+				clusterv1.ManagedByAnnotation: managedByAnnotationValueClusterCAPIOperatorInfraClusterController,
+			},
+		},
+
+		Spec: azurev1.AzureClusterSpec{
+			AzureClusterClassSpec: azurev1.AzureClusterClassSpec{
+				Location:         location,
+				AzureEnvironment: "AzurePublicCloud",
```

Comment:
What do we do for AzureStack? Do we have a check before we create these resources that we are on public cloud and not stack?

Reply:
Maybe a question for @damdo when he's back, I'm not very knowledgeable on AzureStack. There's no checking done around AzureStack in our e2es either. I think the right behaviour is probably to fix the e2es up first, wdyt?